core/src/test/kotlin/org/eln2/data/SetMapMultiMapTest.kt

```diff
@@ -0,0 +1,286 @@
+package org.eln2.data
+
+import org.junit.jupiter.api.Assertions.*
+import org.junit.jupiter.api.Test
+
+// FIXME: abstract this so that the Mutable version below isn't LITERALLY an unmaintainable copy of this
+internal class SetMapMultiMapTest {
+
+	companion object {
+		val TEST_MAP: Map<Int, Int> = mapOf(1 to 2, 2 to 4, 4 to 8, 8 to 16, 16 to 32)
+		val TEST_MULTIMAP: List<Pair<Int, Int>> = listOf(
+			Pair(1, 2),
+			Pair(2, 1), Pair(2, 2),
+			Pair(4, 1), Pair(4, 2), Pair(4, 3), Pair(4, 4),
+			Pair(8, 1), Pair(8, 2), Pair(8, 3), Pair(8, 4), Pair(8, 5), Pair(8, 6), Pair(8, 7), Pair(8, 8)
+		)
+		val TEST_MULTIMAP_KEYS = 4
+		val TEST_MULTIMAP_UNIQUE_VALUES = 8
+		val TEST_NONKEY = 7
+	}
+
+	// FIXME: Kotlin can't seem to understand Pair<K,V> in the parameter below, so concrete types only for now
+	protected fun makeImplementation(iter: List<Pair<Int, Int>>): MultiMap<Int, Int> = SetMapMultiMap(iter.iterator())
+	
+	@Test
+	fun emptySizes() {
+		val map = makeImplementation(emptyList())
+
+		assertEquals(map.entriesSize, 0)
+		assertEquals(map.keyMappingSize, 0)
+		assertEquals(map.uniqueValuesSize, 0)
+	}
+	
+	@Test
+	fun emptyCollections() {
+		val map = makeImplementation(emptyList())
+
+		assertEquals(map.entries.count(), 0)
+		assertEquals(map.keyMapping.count(), 0)
+		assertEquals(map.uniqueValues.size, 0)
+	}
+
+	@Test
+	fun singleMapConsistency() {
+		val map = makeImplementation(TEST_MAP.entries.map { it.toPair() })
+
+		assertEquals(map.entriesSize, TEST_MAP.size)
+		assertEquals(map.entriesSize, map.entries.count())
+		assertEquals(map.keyMappingSize, TEST_MAP.size)
+		assertEquals(map.keyMappingSize, map.keyMapping.count())
+		assertEquals(map.uniqueValuesSize, TEST_MAP.size)
+		assertEquals(map.uniqueValuesSize, map.uniqueValues.size)
+
+		TEST_MAP.entries.forEach { (k, v) ->
+			assert(k in map)
+			val set = map[k]
+			assertEquals(set.size, 1)
+			assertEquals(set.iterator().next(), v)
+		}
+
+		assert(TEST_NONKEY !in map)
+		assert(map[TEST_NONKEY].isEmpty())
+	}
+
+	@Test
+	fun multiMapConsistency() {
+		val map = makeImplementation(TEST_MULTIMAP)
+
+		println("multimap keymappings ${map.keyMapping} entries: ${map.entries} size ${map.entriesSize}")
```

Comment:
Tests shouldn't print anything except if they fail.

Reply:
Sorry, some left-over debug prints. I feel like that could guide someone hunting a regression, though; mind if I leave it there but commented as a "known good debug tool"? (I don't imagine leaving commented-out code in _test cases_ is the most evil thing...)

Reply:
You could also use a try-finally, I think.